### PR TITLE
Moved CordaPersistence and dependent classes into internal package in…

### PR DESCRIPTION
**Pull request by shamsasari**
… node-api. This is so that the doorman can make use of them without needing a compile dependency to node.

### CODE REVIEW

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/CordaPersistence.kt [hunk 1]
```diff
@@ -1,64 +1,75 @@
...
+
 class CordaPersistence(
-        val dataSource: HikariDataSource,
-        private val schemaService: SchemaService,
-        private val identityService: IdentityService,
-        databaseConfig: DatabaseConfig
+        val dataSource: DataSource,
+        databaseConfig: DatabaseConfig,
+        schemas: Set<MappedSchema>,
+        attributeConverters: List<AttributeConverter<*, *>>
```
Comment (josecoll) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/CordaPersistence.kt]: Can we make this parameter optional (? = null)? 
(as may not be needed at all for simple CordaPersistence)

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/CordaPersistence.kt [hunk 2]
```diff
@@ -1,64 +1,75 @@
...
             }
         }
     }
 
     /**
-     * Creates an instance of [DatabaseTransaction], with the given isolation level.
-     * @param isolationLevel isolation level for the transaction. If not specified the default (i.e. provided at the creation time) is used.
+     * Creates an instance of [DatabaseTransaction], with the transaction isolation level specified at the creation time.
      */
-    fun createTransaction(isolationLevel: Int): DatabaseTransaction {
```
Comment (josecoll) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/CordaPersistence.kt]: Good removal - shouldn't be changing the isolation level. 
I think this was added to hack around a Persistent Network Map Cache issue in `updateInfoDB`.

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/CordaPersistence.kt [hunk 3]
```diff
@@ -116,21 +127,8 @@ class CordaPersistence(
...
-fun configureDatabase(dataSourceProperties: Properties, databaseConfig: DatabaseConfig, identityService: IdentityService, schemaService: SchemaService = NodeSchemaService(null)): CordaPersistence {
-    val config = HikariConfig(dataSourceProperties)
-    val dataSource = HikariDataSource(config)
-    val persistence = CordaPersistence.connect(dataSource, schemaService, identityService, databaseConfig)
-    // Check not in read-only mode.
-    persistence.transaction {
-        persistence.dataSource.connection.use {
-            check(!it.metaData.isReadOnly) { "Database should not be readonly." }
-        }
+        (dataSource as? AutoCloseable)?.close()
```
Comment (al-r3) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/CordaPersistence.kt]: This looks a bit odd, why can this be AutoCloseable or not?
Add a comment

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/HibernateConfiguration.kt [hunk 4]
```diff
@@ -18,28 +14,27 @@ import org.hibernate.engine.jdbc.connections.spi.ConnectionProvider
...
 import org.hibernate.type.descriptor.sql.VarbinaryTypeDescriptor
 import java.sql.Connection
 import java.util.concurrent.ConcurrentHashMap
+import javax.persistence.AttributeConverter
 
-class HibernateConfiguration(val schemaService: SchemaService, private val databaseConfig: DatabaseConfig, private val identityService: IdentityService) {
+class HibernateConfiguration(
+        schemas: Set<MappedSchema>,
+        private val databaseConfig: DatabaseConfig,
+        private val attributeConverters: List<AttributeConverter<*, *>>
```
Comment (al-r3) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/HibernateConfiguration.kt]: nit: attributeConverters can be a Collection

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/CordaPersistence.kt (hunk 1 above)
Comment (shamsasari) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/CordaPersistence.kt]: I can default it to an empty list, which is the better thing to do.

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/CordaPersistence.kt (hunk 2 above)
Comment (shamsasari) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/CordaPersistence.kt]: No it's used by the doorman. I think it was a mistake to add it as the isolation level should have been set in the properties. Now that we have `DatabaseConfig` it should be easier.

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/DatabaseTransaction.kt [hunk 5]
```diff
@@ -0,0 +1,61 @@
+package net.corda.nodeapi.internal.persistence
+
+import org.hibernate.Session
+import org.hibernate.Transaction
+import rx.subjects.Subject
+import java.sql.Connection
+import java.util.*
+
+class DatabaseTransaction(
```
Comment (al-r3) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/DatabaseTransaction.kt]: Add javadoc.
Also specify why you need a ThreadLocal

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/HibernateConfiguration.kt [hunk 6]
```diff
@@ -18,28 +14,27 @@ import org.hibernate.engine.jdbc.connections.spi.ConnectionProvider
...
 import org.hibernate.type.descriptor.java.PrimitiveByteArrayTypeDescriptor
 import org.hibernate.type.descriptor.sql.BlobTypeDescriptor
 import org.hibernate.type.descriptor.sql.VarbinaryTypeDescriptor
 import java.sql.Connection
 import java.util.concurrent.ConcurrentHashMap
+import javax.persistence.AttributeConverter
 
-class HibernateConfiguration(val schemaService: SchemaService, private val databaseConfig: DatabaseConfig, private val identityService: IdentityService) {
+class HibernateConfiguration(
+        schemas: Set<MappedSchema>,
```
Comment (josecoll) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/HibernateConfiguration.kt]: Excellent refactoring as this was previously always bootstrapping ALL the schemas defined in `NodeSchemaService` (whether they are required or not).

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/CordaPersistence.kt (hunk 3 above)
Comment (shamsasari) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/CordaPersistence.kt]: Done

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/HibernateConfiguration.kt (hunk 4 above)
Comment (shamsasari) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/HibernateConfiguration.kt]: Done

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/DatabaseTransaction.kt (hunk 5 above)
Comment (shamsasari) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/DatabaseTransaction.kt]: I just moved this class into its own file. Rather not document it when I haven't had the time to study it.

**Comment by shamsasari** [comment]
The `schemaService: SchemaService = NodeSchemaService(null)` declarations are needed now that CordaPeristence doesn't hold a SchemaService. Not sure how else to do it.

Otherwise back to you @josecoll.

**Comment by josecoll** [comment]
As part of this refactor / clean-up, would it be possible to also update `NodeSchemaService` to not take a `cordappLoader: CordappLoader?` constructor argument, but instead a set of schemas:
`Set<MappedSchema>`. Would help immensely with some work I am doing.

**Comment by shamsasari** [comment]
Done what you asked, but had to add a second c'tor for the CordappLoader as it's required in more than one place.

**Comment by josecoll** [comment]
Can you not just remove the CorDapp constructor completely, and just replace the 2 calls to it using the new default constructor and passing in `cordappLoader?.cordapps?.flatMap { it.customSchemas }` ?
(NB: `SchemaOptions` is not used anywhere afaik)

**Comment by shamsasari** [comment]
Moved `cordappLoader?.cordapps?.flatMap { it.customSchemas }` into CordappLoader. So now everything is nice and clean!

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/CordaPersistence.kt (hunk 2 above)
Comment (shamsasari) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/CordaPersistence.kt]: Just spoke with @mkit and he confirmed that we need this as an override for specific cases. In the doorman case we need to use the highest level (serialised?) in certain situations which is not use suitable as a global level. I'll rename some stuff to make this more clear.

**Pull request merged**